File: backend/src/services/ai-ops/batchProcessor.ts
/**
 * BatchProcessor 批处理服务
 * 积攒告警进行批量 AI 分析，优化 API 调用
 *
 * Requirements: 3.1, 3.2, 3.3, 3.7
 * - 3.1: 在 5 秒窗口内积攒告警进行批量分析
 * - 3.2: 批处理窗口过期时，将所有积攒的告警发送给 AI 分析
 * - 3.3: AI 返回分析结果后，分发给对应的告警事件
 * - 3.7: 批次大小超过 20 时，分割成多个批次
 */

import {
  AlertEvent,
  BatchConfig,
  BatchItem,
  IBatchProcessor,
} from '../../types/ai-ops';
import { logger } from '../../utils/logger';
import { aiAnalyzer } from './aiAnalyzer';
import { fingerprintCache } from './fingerprintCache';

// 默认配置
const DEFAULT_CONFIG: BatchConfig = {
  windowMs: 5000,           // 5 秒批处理窗口
  maxBatchSize: 20,         // 最大批次大小
};

export class BatchProcessor implements IBatchProcessor {
  private config: BatchConfig;
  private batch: BatchItem[] = [];
  private batchTimer: NodeJS.Timeout | null = null;
  private running: boolean = false;
  private processing: boolean = false;

  constructor(config?: Partial<BatchConfig>) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    logger.info('BatchProcessor initialized', { config: this.config });
  }

  /**
   * 添加告警到批次，返回 Promise 等待分析结果
   * @param alert 告警事件
   * @returns 分析结果 Promise
   */
  add(alert: AlertEvent): Promise<string> {
    return new Promise((resolve, reject) => {
      const item: BatchItem = { alert, resolve, reject };
      this.batch.push(item);

      logger.debug(`Alert added to batch: ${alert.id}, batch size: ${this.batch.length}`);

      // 如果批次达到最大大小，立即处理
      if (this.batch.length >= this.config.maxBatchSize) {
        logger.info(`Batch size reached max (${this.config.maxBatchSize}), processing immediately`);
        this.processBatch();
      } else if (this.running && !this.batchTimer) {
        // 启动批处理定时器
        this.startBatchTimer();
      }
    });
  }

  /**
   * 立即处理当前批次
   */
  async flush(): Promise<void> {
    if (this.batch.length > 0) {
      await this.processBatch();
    }
  }

  /**
   * 获取待处理数量
   */
  getPendingCount(): number {
    return this.batch.length;
  }

  /**
   * 启动批处理
   */
  start(): void {
    if (this.running) return;
    this.running = true;
    logger.info('BatchProcessor started');
  }

  /**
   * 停止批处理
   */
  stop(): void {
    this.running = false;
    this.stopBatchTimer();
    logger.info('BatchProcessor stopped');
  }

  /**
   * 启动批处理定时器
   */
  private startBatchTimer(): void {
    if (this.batchTimer) return;

    this.batchTimer = setTimeout(() => {
      this.batchTimer = null;
      if (this.batch.length > 0) {
        this.processBatch();
      }
    }, this.config.windowMs);
  }

  /**
   * 停止批处理定时器
   */
  private stopBatchTimer(): void {
    if (this.batchTimer) {
      clearTimeout(this.batchTimer);
      this.batchTimer = null;
    }
  }


  /**
   * 处理当前批次
   */
  private async processBatch(): Promise<void> {
    if (this.processing || this.batch.length === 0) return;

    this.processing = true;
    this.stopBatchTimer();

    // 取出当前批次
    const currentBatch = [...this.batch];
    this.batch = [];

    logger.info(`Processing batch of ${currentBatch.length} alerts`);

    try {
      // 如果批次超过最大大小，分割成多个批次
      const batches = this.splitBatches(currentBatch);

      for (const batch of batches) {
        await this.processSingleBatch(batch);
      }
    } catch (error) {
      logger.error('Batch processing failed:', error);
      // 对所有未处理的项返回错误
      for (const item of currentBatch) {
        item.reject(error instanceof Error ? error : new Error(String(error)));
      }
    } finally {
      this.processing = false;

      // 如果还有待处理的告警，启动新的定时器
      if (this.running && this.batch.length > 0) {
        this.startBatchTimer();
      }
    }
  }

  /**
   * 分割批次
   * @param items 批处理项列表
   * @returns 分割后的批次数组
   */
  private splitBatches(items: BatchItem[]): BatchItem[][] {
    const batches: BatchItem[][] = [];
    for (let i = 0; i < items.length; i += this.config.maxBatchSize) {
      batches.push(items.slice(i, i + this.config.maxBatchSize));
    }
    return batches;
  }

  /**
   * 处理单个批次
   * @param batch 批处理项列表
   */
  private async processSingleBatch(batch: BatchItem[]): Promise<void> {
    logger.debug(`Processing single batch of ${batch.length} alerts`);

    // 构建批量分析请求
    const alertsInfo = batch.map((item, index) => ({
      index,
      id: item.alert.id,
      ruleName: item.alert.ruleName,
      severity: item.alert.severity,
      metric: item.alert.metric,
      currentValue: item.alert.currentValue,
      threshold: item.alert.threshold,
      message: item.alert.message,
    }));

    try {
      // 调用 AI 进行批量分析
      const batchAnalysis = await this.analyzeBatch(alertsInfo);

      // 分发结果给各个告警
      for (let i = 0; i < batch.length; i++) {
        const item = batch[i];
        const analysis = batchAnalysis[i] || this.getDefaultAnalysis(item.alert);
        item.resolve(analysis);
      }

      logger.info(`Batch analysis completed for ${batch.length} alerts`);
    } catch (error) {
      logger.error('Single batch processing failed:', error);
      // 对批次中的所有项返回错误
      for (const item of batch) {
        item.reject(error instanceof Error ? error : new Error(String(error)));
      }
    }
  }

  /**
   * 批量分析告警
   * @param alertsInfo 告警信息列表
   * @returns 分析结果数组
   */
  private async analyzeBatch(alertsInfo: Array<{
    index: number;
    id: string;
    ruleName: string;
    severity: string;
    metric: string;
    currentValue: number;
    threshold: number;
    message: string;
  }>): Promise<string[]> {
    // 构建批量分析提示词
    const prompt = this.buildBatchPrompt(alertsInfo);

    try {
      // 调用 AI 分析
      const result = await aiAnalyzer.analyze({
        type: 'alert',
        context: {
          batchMode: true,
          alerts: alertsInfo,
          prompt,
        },
      });

      // 解析批量分析结果
      return this.parseBatchResult(result.summary, alertsInfo.length);
    } catch (error) {
      logger.warn('AI batch analysis failed, using fallback:', error);
      // 返回默认分析结果
      return alertsInfo.map(info => this.getDefaultAnalysisForInfo(info));
    }
  }

  /**
   * 构建批量分析提示词
   */
  private buildBatchPrompt(alertsInfo: Array<{
    index: number;
    id: string;
    ruleName: string;
    severity: string;
    metric: string;
    currentValue: number;
    threshold: number;
    message: string;
  }>): string {
    const alertsList = alertsInfo.map((info, i) => 
      `[告警 ${i + 1}] ${info.ruleName} (${info.severity}): ${info.message} - 当前值: ${info.currentValue}, 阈值: ${info.threshold}`
    ).join('\n');

    return `请分析以下 ${alertsInfo.length} 个告警事件，为每个告警提供简要分析和建议。

${alertsList}

请按照以下 JSON 格式返回分析结果：
{
  "analyses": [
    {"index": 0, "analysis": "告警1的分析和建议"},
    {"index": 1, "analysis": "告警2的分析和建议"},
    ...
  ]
}`;
  }

  /**
   * 解析批量分析结果
   * @param result AI 返回的结果
   * @param expectedCount 期望的结果数量
   * @returns 分析结果数组
   */
  private parseBatchResult(result: string, expectedCount: number): string[] {
    const analyses: string[] = [];

    try {
      // 尝试解析 JSON 格式
      const jsonMatch = result.match(/\{[\s\S]*"analyses"[\s\S]*\}/);
      if (jsonMatch) {
        const parsed = JSON.parse(jsonMatch[0]);
        if (Array.isArray(parsed.analyses)) {
          for (const item of parsed.analyses) {
            analyses[item.index] = item.analysis;
          }
        }
      }
    } catch {
      // JSON 解析失败，尝试按段落分割
      const sections = result.split(/\[告警\s*\d+\]/);
      for (let i = 1; i < sections.length && i <= expectedCount; i++) {
        analyses[i - 1] = sections[i].trim();
      }
    }

    // 确保返回正确数量的结果
    while (analyses.length < expectedCount) {
      analyses.push('分析结果不可用');
    }

    return analyses;
  }

  /**
   * 获取默认分析结果
   */
  private getDefaultAnalysis(alert: AlertEvent): string {
    return `[${alert.severity}] ${alert.ruleName}: ${alert.message}。建议检查相关配置和系统状态。`;
  }

  /**
   * 获取默认分析结果（基于告警信息）
   */
  private getDefaultAnalysisForInfo(info: {
    ruleName: string;
    severity: string;
    message: string;
  }): string {
    return `[${info.severity}] ${info.ruleName}: ${info.message}。建议检查相关配置和系统状态。`;
  }

  /**
   * 获取配置
   */
  getConfig(): BatchConfig {
    return { ...this.config };
  }

  /**
   * 更新配置
   */
  updateConfig(config: Partial<BatchConfig>): void {
    this.config = { ...this.config, ...config };
    logger.info('BatchProcessor config updated', { config: this.config });
  }
}

// 导出单例实例
export const batchProcessor = new BatchProcessor();

File: opsevo-python/src/opsevo/services/ai_ops/batch_processor.py
"""
BatchProcessor 批处理服务
积攒告警进行批量 AI 分析，优化 API 调用。

- 在窗口期内积攒告警进行批量分析
- 窗口过期时将所有积攒的告警发送给 AI 分析
- AI 返回分析结果后分发给对应的告警事件
- 批次大小超过 max_batch_size 时分割成多个批次
"""

from __future__ import annotations

import asyncio
import json
import re
from dataclasses import dataclass, field
from typing import Any, Callable, Awaitable

import structlog

logger = structlog.get_logger(__name__)


@dataclass
class BatchConfig:
    window_seconds: float = 5.0
    max_batch_size: int = 20


@dataclass
class _BatchItem:
    alert: dict[str, Any]
    future: asyncio.Future[str] = field(default=None)  # type: ignore[assignment]


class BatchProcessor:
    """批处理服务：积攒告警后批量调用 AI 分析。"""

    def __init__(
        self,
        ai_analyzer: Any | None = None,
        config: BatchConfig | None = None,
    ) -> None:
        self._config = config or BatchConfig()
        self._ai_analyzer = ai_analyzer
        self._batch: list[_BatchItem] = []
        self._timer_task: asyncio.Task[None] | None = None
        self._processing = False
        self._running = False
        logger.info("BatchProcessor initialized", config=self._config)

    # ------------------------------------------------------------------
    async def add(self, alert: dict[str, Any]) -> str:
        loop = asyncio.get_running_loop()
        item = _BatchItem(alert=alert, future=loop.create_future())
        self._batch.append(item)

        if len(self._batch) >= self._config.max_batch_size:
            await self._process_batch()
        elif self._running and self._timer_task is None:
            self._start_timer()

        return await item.future

    async def flush(self) -> None:
        if self._batch:
            await self._process_batch()

    @property
    def pending_count(self) -> int:
        return len(self._batch)

    def start(self) -> None:
        if self._running:
            return
        self._running = True
        logger.info("BatchProcessor started")

    def stop(self) -> None:
        self._running = False
        self._cancel_timer()
        logger.info("BatchProcessor stopped")

    # ------------------------------------------------------------------
    # internals
    # ------------------------------------------------------------------
    def _start_timer(self) -> None:
        if self._timer_task is not None:
            return

        async def _tick() -> None:
            await asyncio.sleep(self._config.window_seconds)
            self._timer_task = None
            if self._batch:
                await self._process_batch()

        self._timer_task = asyncio.create_task(_tick())

    def _cancel_timer(self) -> None:
        if self._timer_task is not None:
            self._timer_task.cancel()
            self._timer_task = None

    async def _process_batch(self) -> None:
        if self._processing or not self._batch:
            return
        self._processing = True
        self._cancel_timer()

        current = list(self._batch)
        self._batch.clear()

        try:
            for chunk in self._split(current):
                await self._process_single(chunk)
        except Exception as exc:
            for item in current:
                if not item.future.done():
                    item.future.set_exception(exc)
        finally:
            self._processing = False
            if self._running and self._batch:
                self._start_timer()

    def _split(self, items: list[_BatchItem]) -> list[list[_BatchItem]]:
        sz = self._config.max_batch_size
        return [items[i : i + sz] for i in range(0, len(items), sz)]

    async def _process_single(self, batch: list[_BatchItem]) -> None:
        infos = [
            {
                "index": i,
                "id": it.alert.get("id", ""),
                "ruleName": it.alert.get("ruleName", ""),
                "severity": it.alert.get("severity", ""),
                "metric": it.alert.get("metric", ""),
                "currentValue": it.alert.get("currentValue", 0),
                "threshold": it.alert.get("threshold", 0),
                "message": it.alert.get("message", ""),
            }
            for i, it in enumerate(batch)
        ]

        try:
            analyses = await self._analyze_batch(infos)
            for i, item in enumerate(batch):
                result = analyses[i] if i < len(analyses) else self._default(item.alert)
                if not item.future.done():
                    item.future.set_result(result)
        except Exception as exc:
            for item in batch:
                if not item.future.done():
                    item.future.set_exception(exc)

    async def _analyze_batch(self, infos: list[dict]) -> list[str]:
        prompt = self._build_prompt(infos)
        if self._ai_analyzer is None:
            return [self._default_info(i) for i in infos]
        try:
            result = await self._ai_analyzer.analyze(
                {"type": "alert", "context": {"batchMode": True, "alerts": infos, "prompt": prompt}}
            )
            return self._parse_result(result.get("summary", ""), len(infos))
        except Exception:
            logger.warning("AI batch analysis failed, using fallback")
            return [self._default_info(i) for i in infos]

    @staticmethod
    def _build_prompt(infos: list[dict]) -> str:
        lines = [
            f"[告警 {i+1}] {a['ruleName']} ({a['severity']}): {a['message']} - 当前值: {a['currentValue']}, 阈值: {a['threshold']}"
            for i, a in enumerate(infos)
        ]
        return (
            f"请分析以下 {len(infos)} 个告警事件，为每个告警提供简要分析和建议。\n\n"
            + "\n".join(lines)
            + '\n\n请按照以下 JSON 格式返回分析结果：\n{"analyses": [{"index": 0, "analysis": "..."}]}'
        )

    @staticmethod
    def _parse_result(text: str, expected: int) -> list[str]:
        analyses: list[str | None] = [None] * expected
        try:
            m = re.search(r'\{[\s\S]*"analyses"[\s\S]*\}', text)
            if m:
                parsed = json.loads(m.group(0))
                for item in parsed.get("analyses", []):
                    idx = item.get("index", -1)
                    if 0 <= idx < expected:
                        analyses[idx] = item.get("analysis", "分析结果不可用")
        except (json.JSONDecodeError, KeyError):
            pass
        return [a if a is not None else "分析结果不可用" for a in analyses]

    @staticmethod
    def _default(alert: dict) -> str:
        return f"[{alert.get('severity','')}] {alert.get('ruleName','')}: {alert.get('message','')}。建议检查相关配置和系统状态。"

    @staticmethod
    def _default_info(info: dict) -> str:
        return f"[{info.get('severity','')}] {info.get('ruleName','')}: {info.get('message','')}。建议检查相关配置和系统状态。"
